hazelcast/src/test/java/com/hazelcast/sql/impl/worker/QueryFragmentExecutableTest.java

```diff
@@ -0,0 +1,478 @@
+            };
+
+            OutboundHandler outboundHandler = outboundQueue::add;
+
+            // Prepare executable.
+            QueryId queryId = QueryId.create(UUID.randomUUID());
+            UUID callerId = UUID.randomUUID();
+            int edgeId = 1;
+
+            pool = createPool();
+
+            TestStateCallback stateCallback = new TestStateCallback();
+
+            TestExec exec = new TestExec().setPayload(() -> {
+                while (!inboundQueue.isEmpty()) {
+                    inboundSet.add(inboundQueue.poll());
+                }
+
+                while (!outboundQueue.isEmpty()) {
+                    outboundSet.add(outboundQueue.poll());
+                }
+
+                return IterationResult.FETCHED;
+            });
+
+            QueryFragmentExecutable fragmentExecutable = new QueryFragmentExecutable(
+                stateCallback,
+                Collections.emptyList(),
+                exec,
+                Collections.singletonMap(edgeId, inboundHandler),
+                Collections.singletonMap(edgeId, Collections.singletonMap(callerId, outboundHandler)),
+                pool
+            );
+
+            assertEquals(1, fragmentExecutable.getInboxEdgeIds().size());
+            assertEquals(1, fragmentExecutable.getOutboxEdgeIds().size());
+            assertEquals(edgeId, (int) fragmentExecutable.getInboxEdgeIds().iterator().next());
+            assertEquals(edgeId, (int) fragmentExecutable.getOutboxEdgeIds().iterator().next());
+
+            // Start threads and wait for completion.
+            int operationCount = 10_000;
+            int threadCount = 8;
+
+            AtomicInteger doneOperationCount = new AtomicInteger();
+            CountDownLatch doneLatch = new CountDownLatch(threadCount);
+
+            Runnable run = () -> {
+                try {
+                    while (true) {
+                        int counter = doneOperationCount.getAndIncrement();
+
+                        if (counter >= operationCount) {
+                            break;
+                        }
+
+                        QueryAbstractExchangeOperation operation;
+
+                        if (ThreadLocalRandom.current().nextBoolean()) {
+                            operation = new QueryBatchExchangeOperation(queryId, edgeId, EmptyRowBatch.INSTANCE, false, counter);
+                        } else {
+                            operation = new QueryFlowControlExchangeOperation(queryId, edgeId, counter);
+                        }
+
+                        operation.setCallerId(callerId);
+
+                        fragmentExecutable.addOperation(operation);
+                        fragmentExecutable.schedule();
+                    }
+                } finally {
+                    doneLatch.countDown();
+                }
+            };
+
+            for (int j = 0; j < threadCount; j++) {
+                new Thread(run).start();
+            }
+
+            doneLatch.await();
```

Comment:
You can use `hazelcastTestSupport#assertOpenEventually` which has a timeout to avoid indefinite blocks...

Reply:
The release of this latch is part of the test infrastructure, not testing logic. Asserting test code doesn't seem natural. In any case, the timeout worker will cause thread interrupt anyway in case of bugs in the test itself, so we are safe here.